### PR TITLE
EEA3 and EIA3 support

### PR DESCRIPTION
**Pull request by davidrupprecht**
Hi all,

I have added EIA3 and EEA3 support to srsLTE. E*A3 is based on the ZUC cipher and optional in 3GPP LTE networks. Some networks and phones do support it and it might be useful to have it in srsLTE as well. For the cipher function, I use the reference implantation by the specs. I have added all test cases for encryption and integrity protection. 

If we pull this branch in the master, we should also adjust the security algorithm selection of #310 and also the security selection algorithm in the MME. I can take care of this. 

Further, I noticed that we have two snow_3g implementation in the code. I would merge the for those two versions and clean up the file liblte_security.cc a bit, if would do not mind. 

Best regards,
David

### CODE REVIEW

**Comment by andrepuschmann** [comment]
Hey David, thanks for that. Much appreciated. We are currently preparing the new release and have refactored the EPC quite a bit for that. I guess there will be quite a few conflicts with that PR in the current form. Why don't we wait for this for the next release and you rebase the PR then accordingly?

**Comment by davidrupprecht** [comment]
Hi,

yes, I can rebase on the next release and then reissue the pull request. Are you pulling #310 in? 

Best regards,
David

**Comment by andrepuschmann** [comment]
Yes. #310 is currently being checked and pulled in. I suggest you wait for that and then rebase.

**Comment by davidrupprecht** [comment]
Great, that is why I asked for it :)

**Comment by andrepuschmann** [comment]
Sure. Also check the nas_test ;-)

**Comment by davidrupprecht** [comment]
Yeah, just saw it. I have noticed, that when you build srsLTE in release mode the asserts are not triggered. That is why the tests passed on my PC.

**Comment by andrepuschmann** [comment]
So you always compile in release and never debug? Brave man ;-) But seriously, those tests should really not use assert to also trigger in release mode.

**Comment by davidrupprecht** [comment]
Sorry, my fault. It should have triggered. You are right, I will change the assert statements in the encryption test cases.

**Comment by davidrupprecht** [comment]
Rebased 19.06 version

File: lib/src/common/liblte_security.cc [hunk 1]
```diff
@@ -807,6 +808,92 @@ LIBLTE_ERROR_ENUM liblte_security_128_eia2(
...
+        // Generate keystream
+        int N = msg_len + 64;
+        int L = (N + 31) / 32;
+
+        ks = (uint32*)calloc(L, sizeof(uint32));
+
+        zuc_generate_keystream(&zuc_state, L, ks);
+
+        uint32_t T = 0;
+        for (i = 0; i < msg_len; i++) {
```
Comment (andrepuschmann) [inline lib/src/common/liblte_security.cc]: this gives:
```[ 28%] Building CXX object lib/src/common/CMakeFiles/srslte_common.dir/buffer_pool.cc.o
[ 29%] Building CXX object lib/src/common/CMakeFiles/srslte_common.dir/liblte_security.cc.o
/home/travis/build/srsLTE/srsLTE/lib/src/common/liblte_security.cc: In function ‘LIBLTE_ERROR_ENUM liblte_security_128_eia3(uint8*, uint32, uint8, uint8, uint8*, uint32, uint8*)’:
/home/travis/build/srsLTE/srsLTE/lib/src/common/liblte_security.cc:877:23: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
         for (i = 0; i < msg_len; i++) {
                       ^
[ 29%] Building CXX object lib/src/common/CMakeFiles/srslte_common.dir/log_filter.cc.o
```

**Comment by andrepuschmann** [comment]
We've merged that privately after rebasing with next. Will be part of the new release. Thanks again for that.

**Pull request closed without merging**